In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [2]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

def split_scalar(indep_x, dep_Y):
    # Split the data into training and testing sets
    x_train, x_test, y_train, y_test = train_test_split(indep_x, dep_Y, test_size=0.30, random_state=0)
    sc = StandardScaler() 
    # Scale the training data
    x_train = sc.fit_transform(x_train)
    # Scale the testing data using the scaler "sc"
    x_test = sc.transform(x_test)
    return x_train, x_test, y_train, y_test

In [3]:
def apply_pca(indep_x, n_components=2):
    pca = PCA(n_components=n_components)
    indep_x_pca = pca.fit_transform(indep_x)
    return indep_x_pca

In [4]:
def logistic(X_train, y_train, X_test, y_test):
    classifier = LogisticRegression(random_state=0)
    classifier.fit(X_train, y_train)
    return cm_prediction(classifier, X_test, y_test)


In [5]:
def cm_prediction(classifier, X_test, y_test):
    y_pred = classifier.predict(X_test)
    cm = confusion_matrix(y_test, y_pred)
    accuracy = accuracy_score(y_test, y_pred)
    report = classification_report(y_test, y_pred)
    return classifier, accuracy, report, cm


In [6]:
def svm_linear(X_train, y_train, X_test, y_test):
    classifier = SVC(kernel='linear', random_state=0)
    classifier.fit(X_train, y_train)
    return cm_prediction(classifier, X_test, y_test)

In [7]:
def svm_NL(X_train, y_train, X_test, y_test):
    classifier = SVC(kernel='rbf', random_state=0)
    classifier.fit(X_train, y_train)
    return cm_prediction(classifier, X_test, y_test)                        

In [8]:
def Navie(X_train, y_train, X_test, y_test):
    classifier = GaussianNB()
    classifier.fit(X_train, y_train)
    return cm_prediction(classifier, X_test, y_test)

In [9]:
def knn(X_train, y_train, X_test, y_test):
    classifier = KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=2)
    classifier.fit(X_train, y_train)
    return cm_prediction(classifier, X_test, y_test)

In [10]:
def Decision(X_train, y_train, X_test, y_test):
    classifier = DecisionTreeClassifier(criterion='entropy', random_state=0)
    classifier.fit(X_train, y_train)
    return cm_prediction(classifier, X_test, y_test)

In [11]:
def random(X_train, y_train, X_test, y_test):
    classifier = RandomForestClassifier(n_estimators=10, criterion='entropy', random_state=0)
    classifier.fit(X_train, y_train)
    return cm_prediction(classifier, X_test, y_test)

In [12]:
def selectk_Classification(acclog, accsvml, accsvmnl, accknn, accnav, accdes, accrf):
    dataframe = pd.DataFrame(index=['PCA'], columns=['Logistic', 'SVMl', 'SVMnl', 'KNN', 'Navie', 'Decision', 'Random'])
    for number, idx in enumerate(dataframe.index):
        dataframe['Logistic'][idx] = acclog[number]
        dataframe['SVMl'][idx] = accsvml[number]
        dataframe['SVMnl'][idx] = accsvmnl[number]
        dataframe['KNN'][idx] = accknn[number]
        dataframe['Navie'][idx] = accnav[number]
        dataframe['Decision'][idx] = accdes[number]
        dataframe['Random'][idx] = accrf[number]
    return dataframe

In [14]:
# Load dataset
dataset1=pd.read_csv("Car_Rating data.csv",index_col=None)
X = dataset1.iloc[:, 0:13].values
y = dataset1.iloc[:, -1].values

In [15]:
df1= pd.get_dummies(dataset1, drop_first=True)
df1

Index  Overall Rating  Exterior  Interior  Ride Quality  Airbags  \
0        0             4.4       4.4       4.4           4.4       12   
1        1             4.2       4.0       3.8           4.2        8   
2        2             3.5       2.7       3.0           2.3        2   
3        3             4.4       4.4       4.4           4.4        0   
4        4             4.2       4.0       3.8           4.2        4   
..     ...             ...       ...       ...           ...      ...   
167    167             0.0       NaN       NaN           NaN        4   
168    168             0.0       NaN       NaN           NaN        8   
169    169             0.0       NaN       NaN           NaN        4   
170    170             0.0       NaN       NaN           NaN       12   
171    171             0.0       NaN       NaN           NaN        0   

     Price.1  Car_Audi A4  Car_Audi Q3  Car_BMW 2 Series Gran Coupe  ...  \
0      13328        False        False                        False  ...   
1      16621        False        False                        False  ...   
2       8467        False        False                        False  ...   
3       3607        False        False                        False  ...   
4      11726        False        False                        False  ...   
..       ...          ...          ...                          ...  ...   
167    20005        False        False                        False  ...   
168    36110        False        False                        False  ...   
169     7840        False        False                        False  ...   
170      470        False        False                        False  ...   
171     1300        False        False                        False  ...   

     Color_Red  Color_Silver  Color_Sky blue  Color_White  Color_Yellow  \
0        False          True           False        False         False   
1        False         False           False        False         False   
2        False         False           False        False         False   
3        False         False           False         True         False   
4        False          True           False        False         False   
..         ...           ...             ...          ...           ...   
167      False         False           False         True         False   
168      False          True           False        False         False   
169      False         False           False        False         False   
170      False         False           False        False         False   
171      False         False           False        False         False   

     Fuel type_Diesel  Fuel type_Hybrid  Fuel type_LPG  Fuel type_Petrol  \
0               False              True          False             False   
1               False             False          False              True   
2               False             False          False              True   
3               False              True          False             False   
4               False             False          False              True   
..                ...               ...            ...               ...   
167             False             False          False              True   
168             False             False          False              True   
169             False             False          False              True   
170             False             False          False              True   
171             False             False          False              True   

     Fuel type_Plug-in Hybrid  
0                       False  
1                       False  
2                       False  
3                       False  
4                       False  
..                        ...  
167                     False  
168                     False  
169                     False  
170                     False  
171                     False  

[172 rows x 241 

In [16]:
print(dataset1.columns)

Index(['Index', 'Car', 'Price', 'Overall Rating', 'Exterior', 'Interior',
       'Ride Quality', 'Color', 'Airbags', 'Fuel type', 'Price.1'],
      dtype='object')


In [17]:
df1["Exterior"].fillna(0, inplace=True)
df1["Interior"].fillna(0, inplace=True)
df1["Ride Quality"].fillna(0, inplace=True)

In [18]:
# Assuming you have already preprocessed the data and have independent and dependent variables
indep_x=df1.drop('Price.1',axis=1)
dep_Y=df1['Price.1']

In [35]:
pca = apply_pca(indep_x, n_components=1) # Adjusted call without dep_Y

In [36]:
pca = apply_pca(indep_x, n_components=1)
acclog=[]
accsvml=[]
accsvmnl=[]
accknn=[]
accnav=[]
accdes=[]
accrf=[]

In [37]:
x_train,x_test,y_train,y_test = split_scalar(pca,dep_Y)
classifier, accuracy, report, cm = logistic(x_train, y_train, x_test, y_test)
acclog.append(accuracy)

classifier, accuracy, report, cm = svm_linear(x_train, y_train, x_test, y_test)
accsvml.append(accuracy)

classifier, accuracy, report, cm = svm_NL(x_train, y_train, x_test, y_test)
accsvmnl.append(accuracy)

classifier, accuracy, report, cm = knn(x_train, y_train, x_test, y_test)
accknn.append(accuracy)

classifier, accuracy, report, cm = Navie(x_train, y_train, x_test, y_test)
accnav.append(accuracy)

classifier, accuracy, report, cm = Decision(x_train, y_train, x_test, y_test)
accdes.append(accuracy)

classifier, accuracy, report, cm = random(x_train, y_train, x_test, y_test)
accrf.append(accuracy)

# Get results
result = selectk_Classification(acclog, accsvml, accsvmnl, accknn, accnav, accdes, accrf)

C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_cla

In [22]:
result
#5

Logistic      SVMl SVMnl       KNN     Navie Decision Random
PCA      0.0  0.019231   0.0  0.038462  0.038462      0.0    0.0

# KNN and Naive Bayes show the highest performance with a score of 0.038462.
# SVM (linear) has a performance score of 0.019231.
# Logistic, SVM (nonlinear), Decision Tree, and Random Forest have a score of 0.0.

In [26]:
result
#4

Logistic      SVMl SVMnl       KNN Navie Decision Random
PCA      0.0  0.038462   0.0  0.038462   0.0      0.0    0.0

# KNN and SVM (linear) show the highest performance with a score of 0.038462.
# Logistic, SVM (nonlinear), Naive Bayes, Decision Tree, and Random Forest have a score of 0.0.

In [30]:
result
#3

Logistic      SVMl SVMnl       KNN     Navie Decision Random
PCA  0.019231  0.019231   0.0  0.038462  0.057692      0.0    0.0

# Naive Bayes shows the highest performance with a score of 0.057692.
# KNN follows with a score of 0.038462.
# Logistic and SVM (linear) have scores of 0.019231.
# SVM (nonlinear), Decision Tree, and Random Forest have a score of 0.0.

In [34]:
result
#2

Logistic      SVMl     SVMnl       KNN     Navie Decision Random
PCA      0.0  0.019231  0.019231  0.019231  0.096154      0.0    0.0

# Naive Bayes shows the highest performance with a score of 0.096154.
# SVM (linear), SVM (nonlinear), and KNN each have a score of 0.019231.
# Logistic, Decision Tree, and Random Forest have a score of 0.0.

In [38]:
result
#1

Logistic SVMl SVMnl       KNN     Navie Decision Random
PCA      0.0  0.0   0.0  0.019231  0.019231      0.0    0.0

# KNN and Naive Bayes show the highest performance with a score of 0.019231.
# Logistic, SVM (linear), SVM (nonlinear), Decision Tree, and Random Forest have a score of 0.0.